pkg/core/mem_pool.go

```diff
@@ -0,0 +1,250 @@
+			return -1
+		}
+	}
+
+	// Fees sorted ascending
+	pFPB := p.fee.FeePerByte(p.txn)
+	otherFPB := p.fee.FeePerByte(otherP.txn)
+	if ret := pFPB.CompareTo(otherFPB); ret != 0 {
+		return ret
+	}
+
+	pNF := p.fee.NetworkFee(p.txn)
+	otherNF := p.fee.NetworkFee(otherP.txn)
+	if ret := pNF.CompareTo(otherNF); ret != 0 {
+		return ret
+	}
+
+	// Transaction hash sorted descending
+	return otherP.txn.Hash().CompareTo(p.txn.Hash())
+}
+
+// Count returns the total number of uncofirm transactions.
+func (mp MemPool) Count() int {
+	return len(mp.unsortedTxn) + len(mp.unverifiedTxn)
+}
+
+// ContainsKey checks if a transactions hash is in the MemPool.
+func (mp MemPool) ContainsKey(hash util.Uint256) bool {
+	if _, ok := mp.unsortedTxn[hash]; ok {
+		return true
+	}
+
+	if _, ok := mp.unverifiedTxn[hash]; ok {
+		return true
+	}
+
+	return false
+}
+
+// TryAdd try to add the PoolItem to the MemPool.
+func (mp MemPool) TryAdd(hash util.Uint256, pItem *PoolItem) bool {
+	var pool PoolItems
+
+	if _, ok := mp.unsortedTxn[hash]; ok {
+		return false
+	}
+	mp.unsortedTxn[hash] = pItem
+
+	if pItem.fee.IsLowPriority(pItem.txn) {
+		pool = mp.sortedLowPrioTxn
+	} else {
+		pool = mp.sortedHighPrioTxn
+	}
+	pool = append(pool, pItem)
+	sort.Sort(pool)
+
+	if mp.Count() > mp.capacity {
+
+	}
+
+	_, ok := mp.unsortedTxn[hash]
+	return ok
+}
+
+// RemoveOverCapacity removes ...
+func (mp MemPool) RemoveOverCapacity(pool map[util.Uint256]PoolItem, time time.Time) {
+	for mp.Count()-mp.capacity > 0 {
+		if minItem, argPosition := getLowestFeeTransaction(mp.sortedLowPrioTxn, mp.unverifiedSortedLowPrioTxn); minItem != nil {
+			if argPosition == 1 {
+				// minItem belongs to the mp.sortedLowPrioTxn slice.
+				// The corresponding unsorted pool is is mp.unsortedTxn.
+				delete(mp.unsortedTxn, minItem.txn.Hash())
+				mp.sortedLowPrioTxn = append(mp.sortedLowPrioTxn[:0], mp.sortedLowPrioTxn[1:]...)
+			} else {
+				// minItem belongs to the mp.unverifiedSortedLowPrioTxn slice.
+				// The corresponding unsorted pool is is mp.unverifiedTxn.
+				delete(mp.unverifiedTxn, minItem.txn.Hash())
+				mp.unverifiedSortedLowPrioTxn = append(mp.unverifiedSortedLowPrioTxn[:0], mp.unverifiedSortedLowPrioTxn[1:]...)
+
+			}
+		} else if minItem, argPosition := getLowestFeeTransaction(mp.sortedHighPrioTxn, mp.unverifiedSortedHighPrioTxn); minItem != nil {
+			if argPosition == 1 {
+				// minItem belongs to the mp.sortedHighPrioTxn slice.
+				// The corresponding unsorted pool is is mp.unsortedTxn.
+				delete(mp.unsortedTxn, minItem.txn.Hash())
+				mp.sortedHighPrioTxn = append(mp.sortedHighPrioTxn[:0], mp.sortedHighPrioTxn[1:]...)
+			} else {
+				// minItem belongs to the mp.unverifiedSortedHighPrioTxn slice.
+				// The corresponding unsorted pool is is mp.unverifiedTxn.
+				delete(mp.unverifiedTxn, minItem.txn.Hash())
+				mp.unverifiedSortedHighPrioTxn = append(mp.unverifiedSortedHighPrioTxn[:0], mp.unverifiedSortedHighPrioTxn[1:]...)
```

Comment:
- https://github.com/CityOfZion/neo-go/pull/174/files#diff-8d49ac90d65173c888ff0f35ec16762eR128
- https://github.com/CityOfZion/neo-go/pull/174/files#diff-8d49ac90d65173c888ff0f35ec16762eR133
- https://github.com/CityOfZion/neo-go/pull/174/files#diff-8d49ac90d65173c888ff0f35ec16762eR141

it won't work like you expect, see example:
1) https://play.golang.org/p/C7-ZCgY6Pnf (like in your code)
2) https://play.golang.org/p/Ncctt9Qg0bD (fixed code)

Reply:
Thanks! This is really a nice error that you found it and fix it! :+1:

Reply:
sync.Mutex -> sync.RWMutex

And use read/write lock

Reply:
:+1: